framework/audio/engine/internal/audioengine.cpp

```diff
@@ -167,12 +228,76 @@ void AudioEngine::setPlaySpeed(handle h, float speed)
     m_sl->engine.setRelativePlaySpeed(h, speed);
 }
 
-mu::async::Notification AudioEngine::playCallbackCalled() const
+void AudioEngine::swapPlayContext(handle h, Context& ctx)
 {
-    return m_playCallbackCalled;
+    HandleMeta& m = meta(h);
+    IF_ASSERT_FAILED(m.isValid()) {
+        return;
+    }
+
+    m.playContext.swap(ctx);
+    if (m.playContextRequested) {
+        m.playContextChanged.send(m.playContext);
+    }
+
+    if (m.playContext.hasVal(CtxKey::InstanceDestroyed)) {
+        onStop(h);
+    }
 }
 
-void AudioEngine::onPlayCallbackCalled()
+mu::async::Channel<Context> AudioEngine::playContextChanged(handle h) const
 {
-    m_playCallbackCalled.notify();
+    HandleMeta& m = const_cast<AudioEngine*>(this)->meta(h);
+    IF_ASSERT_FAILED(m.isValid()) {
+        return mu::async::Channel<Context>();
+    }
+    m.playContextRequested = true;
+    return m.playContextChanged;
+}
+
+AudioEngine::HandleMeta& AudioEngine::pushMeta(handle h)
+{
+    std::lock_guard<std::mutex> lock(m_metasMutex);
+    auto it = m_handleMetas.find(h);
+    IF_ASSERT_FAILED(it == m_handleMetas.end()) {
```

Comment:
So if we already have the meta for this handle we'll fail with an assert? Is this really a problem that needs to be limited to assertion?

Reply:
I think yes.
If the meta already exists, it means that it was not removed when the handle was destroyed.
It is possible that the value of the handle coincided (there is a guarantee that all the handles are unique at the same time, that is, there is just a pull of the handles), which means that the meta will contain irrelevant data.